ledger-api/grpc-definitions/com/daml/ledger/api/v1/error.proto

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+// Rules on error reporting over the Ledger API
+// ********************************************
+//
+// 1. We use gRPC status codes to report synchronous success and failures
+//    coming from the underlying gRPC protocol or the API-level request processing.
+//    Asynchronous API-level errors are reported over a completion stream like the command completion stream,
+//    which themselves include a ``google.rpc.Status`` message to indicate the status.
+//    The status code ``OK`` denotes success.
+//
+// 2. Error reporting includes relevant error details in the ``details`` field of the status message,
+//    following the gRPC rich error model as described at https://cloud.google.com/apis/design/errors#error_model.
+//    Error details use the `google.rpc.ErrorInfo` message defined at
+//    https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L112.
+//    Its ``reason`` field includes an alphanumeric error code.
+//    Further details on the error and information for subsequent API calls may be included in the ``metadata`` field.
+//    The Ledger API RPC documentation may document further metadata of the error.
+//
+// 3. RPC requests can specify that the API-level error details of the RPC response shall be JSON-encoded in the error message
```

Comment:
What data points are you basing the need for this feature for? I’m aware of http trailers causing issues (or more generally) but you’re not avoiding trailersly. You’re just a bit more robust against things meddling with the trailer. Do we have a client that uses something like this? Or is there a common middleware that does this?

Reply:
I'm not aware of a client that requested this. I've just seen that [grpc-web](https://github.com/grpc/grpc-web) throws away the error details. And some gRPC libraries don't support the error details out of the box; the Java version is fine, but the native Scala bindings don't.

Reply:
Makes sense, afaict grpc-web does actually expose it, it just seems awkward to use it, e.g., https://github.com/shumbo/grpc-web-error-details. I assume by Scala you mean akka-grpc?

Assuming we can fix our own Scala uses, we could probably get away with not having this given that we only support the Java bindings externally and add it later if required but probably simpler to just add it directly.

Reply:
With Scala, I meant what `scalapb` generates. Errors are reported as `StatusRuntimeException` exceptions and if you convert them back into a Status object with `io.grpc.Status#fromThrowable`, the details are gone.

Reply:
If we could avoid adding this feature that would likely save us engineering effort.

So we could also consider doing the PoC's without this feature available, and checking the Dev UX before adding it.